src/cargo/ops/cargo_compile.rs

```diff
@@ -191,6 +195,19 @@ pub fn compile_pkg<'a>(package: &Package,
     let targets = target_with_args.as_ref().map(|&(t, ref p)| vec![(t, p)])
                                            .unwrap_or(targets);
 
+    let mut target_with_rustdoc = None;
+    if let Some(args) = *extra_rustdoc_args {
+        let mut target_with_rustdoc_inner = Vec::new();
+        for &(target, profile) in &targets {
+            let mut profile = profile.clone();
+            profile.rustdoc_args = Some(args.to_vec());
+            target_with_rustdoc_inner.push((target, profile));
+        }
+        target_with_rustdoc = Some(target_with_rustdoc_inner);
+    };
+
+    let targets = target_with_rustdoc.as_ref().map_or(targets,
+                                             |o| o.into_iter().map(|&(t, ref p)| (t, p)).collect());
```

Comment:
This should have the same logic as `cargo rustc` where multiple targets are rejected (and some form of filter is needed)